src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -754,26 +754,29 @@ public void EndBuild()
                 ErrorIfState(BuildManagerState.Idle, "NoBuildInProgress");
                 VerifyStateInternal(BuildManagerState.Building);
 
-                // If there are any submissions which never started, remove them now.
-                var submissionsToCheck = new List<BuildSubmission>(_buildSubmissions.Values);
-                foreach (BuildSubmission submission in submissionsToCheck)
-                {
-                    CheckSubmissionCompletenessAndRemove(submission);
-                }
-
-                var graphSubmissionsToCheck = new List<GraphBuildSubmission>(_graphBuildSubmissions.Values);
-                foreach (GraphBuildSubmission submission in graphSubmissionsToCheck)
-                {
-                    CheckSubmissionCompletenessAndRemove(submission);
-                }
-
                 _buildManagerState = BuildManagerState.WaitingForBuildToComplete;
             }
 
-            ILoggingService loggingService = ((IBuildComponentHost)this).LoggingService;
+            var exceptionsThrownInEndBuild = false;
 
             try
             {
+                lock (_syncLock)
+                {
+                    // If there are any submissions which never started, remove them now.
+                    var submissionsToCheck = new List<BuildSubmission>(_buildSubmissions.Values);
```

Comment:
Why do you need to allocate a new list here and below rather than just iterating over _buildSubmissions.Values?

Reply:
This is just existing code from above that I moved into the try/catch to fail the build in case any exception gets thrown from it, it's not added by me :)
But if I had to guess why, I'd guess that it's cloning the list to protect itself from potential concurrent modifications of the list. Since the access is locked by `_syncLock`, it's either useless copying, or the list can be modified outside the lock. On the other hand there aren't many entries in these lists so it's a not a perf problem.